### DISCUSSION TITLE
Default reasoning

### DISCUSSION THREAD

**Discussion by crapo** [SemanticApplicationDesignLanguage/sadl#623] [Show and tell]
@AbhaMoitra , @agabaldon , @weisenje : I'm looking at implementing default reasoning. There was a primitive version of this in SADL V2  when using the GE-proprietary Jena reasoner. It had several deficiencies, including the way that reasoning was done. In revisiting the topic, I am inviting input. (That's what you're paying me for so you might as well give me some feedback :-)

Consider this model:
![image](https://user-images.githubusercontent.com/1186278/109207877-9fc16f80-7777-11eb-8ead-cc4027dc690b.png)

The previously supported grammar is illustrated on line 16. Lines 17 and 18 show default values with property chains, supported by the SADL V3 grammar. I think that they are also much more potentially useful. Without such property chains it would be much more difficult to say that he default unit for the length of a football field is yds. 

The next step is to implement a reasoner that will handle these default values efficiently. I'm inviting comment on the topic.

**Comment by AbhaMoitra** [comment]
This looks fine. Since you are using "has level 1 default", do you have an illustrative example that uses "level 2" default?

**Comment by agabaldon** [comment]
Any chance we can start experimenting with defaults in the near term?
Is there any draft documentation for defaults?

**Reply by crapo** [reply]
The reasoner/translator pair that supports default values and staged rules is implemented and working but not well-tested. If you have any good examples that can be used in the documentation that would be helpful. I will create a PR to merge the augmented reasoner/translator as soon as all of John's changes are merged. I think I had asked you and @AbhaMoitra to please provide me the default values example(s) from the TestSadlIdeForGE (or something like that) project on OpenGe...

**Reply by agabaldon** [reply]
I uploaded them to sald3/Examples/Defaults directory
